test/Microsoft.AspNetCore.Server.KestrelTests/BadHttpRequestTests.cs

```diff
@@ -1,72 +1,119 @@
+        [InlineData("GET / HTTP/1.1 \r")]
+        [InlineData("GET / HTTP/1.1a\r")]
+        [InlineData("GET / HTTP/1.0\n\r")]
+        [InlineData("GET / HTTP/1.2\r")]
+        [InlineData("GET / HTTP/3.0\r")]
+        [InlineData("GET / H\r")]
+        [InlineData("GET / HTTP/1.\r")]
+        [InlineData("GET / hello\r")]
+        [InlineData("GET / 8charact\r")]
+        // Missing LF
+        [InlineData("GET / HTTP/1.0\rA")]
+        public async Task TestBadRequestLines(string request)
         {
             using (var server = new TestServer(context => { return Task.FromResult(0); }))
             {
                 using (var connection = new TestConnection(server.Port))
                 {
-                    var receiveTask = Task.Run(async () =>
-                    {
-                        await connection.Receive(
-                            "HTTP/1.1 400 Bad Request",
-                            "");
-                        await connection.ReceiveStartsWith("Date: ");
-                        await connection.ReceiveForcedEnd(
-                            "Content-Length: 0",
-                            "Server: Kestrel",
-                            "",
-                            "");
-                    });
-
-                    try
-                    {
-                        await connection.SendEnd(request).ConfigureAwait(false);
-                    }
-                    catch (Exception)
-                    {
-                        // TestConnection.SendEnd will start throwing while sending characters
-                        // in cases where the server rejects the request as soon as it
-                        // determines the request line is malformed, even though there
-                        // are more characters following.
-                    }
+                    await connection.SendEnd(request);
+                    await ReceiveBadRequestResponse(connection, request == "GET / HTTP/1.1\r");
+                }
+            }
+        }
 
-                    await receiveTask;
+        [Theory]
+        [InlineData(" ")]
+        [InlineData("GET  ")]
+        [InlineData("GET / HTTP/1.2\r")]
+        [InlineData("GET / HTTP/1.0\rA")]
+        public async Task ServerClosesConnectionAsSoonAsBadRequestLineIsDetected(string request)
+        {
+            using (var server = new TestServer(context => { return Task.FromResult(0); }))
+            {
+                using (var connection = new TestConnection(server.Port))
+                {
+                    await connection.Send(request);
+                    await ReceiveBadRequestResponse(connection, false);
                 }
             }
         }
+
+        private async Task ReceiveBadRequestResponse(TestConnection connection, bool expectedConnectionCloseHeader)
+        {
+            await connection.Receive(
+                "HTTP/1.1 400 Bad Request",
+                "");
+
+            if (expectedConnectionCloseHeader)
```

Comment:
The RFC only specifies that 400s should be accompanied by a `Connection: close` in situations dealing with transfer encoding (http://tools.ietf.org/html/rfc7230#section-3.3.3). It doesn't specify that it should remain open otherwise, but I'd say keeping it open is more HTTP/1.1-ish.

So now I'm going against what I said before :grin: We *could* change the code to indicate in calls to `Frame.RejectRequest()` whether to keep the connection open or not.